client/blb/client.go

```diff
@@ -1127,109 +1126,123 @@ func (cli *Client) readOneTractReplicated(
+		// 2) Start another goroutine to issue backup requests. Each subsequent request is spawned after the
+		// configured backup time.After delay time expires.
+		// 3) Block until read tractResults are available from the goroutines processing the reads. In the result
+		// collection phase, we mark which hosts reutrned RPC or other errors so we don't re-send reads in the fallback
+		// phase.
+		readFuncBackup := func(wg *sync.WaitGroup, n int, thisBLoc []byte) {
+			defer wg.Done()
+			host := tract.Hosts[order[n]]
 			if host == "" {
 				log.V(1).Infof("read %s from tsid %d: no host", tract.Tract, tract.TSIDs[n])
-				ch <- tractResult{0, 0, err, badVersionHost}
+				ch <- tractResultRepl{
+					tractResult: tractResult{0, 0, err, badVersionHost},
+					thisB:       thisBLoc,
+				}
 				return
 			}
 
-			// TODO(eric): fill in otherHosts when ts-ts cancellation is done.
-			read, err := cli.tractservers.ReadInto(ctx, host, nil, reqID, tract.Tract, tract.Version, thisB, thisOffset)
-			if err == core.ErrVersionMismatch {
-				badVersionHost = host
-			}
-			if err != core.NoError && err != core.ErrEOF {
-				// TODO(eric): redo bad TS reporting mechanism.
-				ch <- tractResult{0, 0, err, badVersionHost}
-				return
-			}
-			log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
-			// In case of short read, i.e., read < len(thisB), we should
-			// pad the result with 0's. This doesn't need to be done
-			// explicitly if 'thisB' given to us starts to be empty.
-			// However, just in case it's not, we will ensure that by
-			// overwritting untouched bytes with 0's.
-			for i := read; i < len(thisB); i++ {
-				thisB[i] = 0
+			result := readFunc(host, n, thisBLoc)
+
+			ch <- tractResultRepl{
+				tractResult: result,
+				badReadHost: host,
+				thisB:       thisBLoc,
 			}
-			ch <- tractResult{len(thisB), read, err, badVersionHost}
-			return
 		}
 
-		// Backup request logic involves the following:
-		// 1) Send the first read to the first host (item 0 available from the orderCh) in a goroutine.
-		// 2) Start another goroutine to issue backup requests. Each subsequent request is spawned after the
-		// configured backup time.After delay time expires.
-		// 3) Block until the first read is returned from the goroutines servicing the reads. Note that if the
-		// the request returned fails, we accept that and continume on to the fallback phase for sequential reads.
-		// We could optimize this later and scan for the first successful backup request, but that may not be
-		// worth the latency savings.
+		var wg sync.WaitGroup
+		nb := min(maxNumBackups+1, len(order))
+		wg.Add(nb)
 		go func() {
-			nb := min(maxNumBackups+1, len(order))
-			for i := 0; i < nb; i++ {
+			for n := 0; n < nb; n++ {
 				select {
-				case <-cli.backupReadState.backupDelayFunc(time.Duration(i) * delay):
-					go readFunc(order[i])
+				case <-cli.backupReadState.delayFunc(time.Duration(n) * delay):
```

Comment:
This structure feels kind of confusing. There's one goroutine that does
```
sleep 0
spawn goroutine for read
sleep 2
spawn goroutine for read
sleep 4
spawn goroutine for read
...
```

So first of all, if there are two backup requests, the second will be delayed by 6ms instead of (I think) the intended 4ms.

Second, the waitgroup handling looks wrong: if the context is cancelled and some requests never get kicked off, the wg will block forever.

It might be cleaner to have just one "level" of goroutines: just start a bunch of them, have one do {sleep 0; read}, have the next do {sleep 2; read}, the third do {sleep 4ms; read}, etc. (with proper context+cancellation handling of course).